### ISSUE TITLE
systemd's out-of-memory killer (`oomd`) is turned on, despite kernel OoM killer (`mglru`) also being enabled on kernels past 6.2 (NixOS 24.05+)

### ISSUE THREAD

**Issue by ahydronous** [NixOS/nixpkgs#338175]
### Describe the bug
From nixOS 24.05 onward, there are two Out-of-Memory killers turned on, the userspace implementation via `systemd-oomd` and the kernel implementation via `mglru`.

### Steps To Reproduce
1. `systemd.oomd.enable` is `true` by default
2. `cat /sys/kernel/mm/lru_gen/enabled` returns `0x0007`, which means enabled

### Expected behavior
Either systemd's or the kernel's OoM killer should be disabled, as running two of them concurrently is counterproductive.
MGLRU is widely considered to be better, so probably "systemd-oomd" should be turned off.
This can be achieved by setting `systemd.oomd.enable` to `false` by default, for nixOS versions past 23.11.

### Notify maintainers
@lf- 

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by K900** [comment]
This is perfectly normal, and not at all counterproductive.

**Comment by ahydronous** [comment]
@mweinelt 
Err.. why is this closed without explanation?

`systemd--oomd`
- works via `cgroupv2` and has a tendency to kill process groups instead of singular runaway processes, which negates the more precise way MGLRU does it
- systemd-oomd` runs in userspace, it can be starved of memory and not make progress
- It uses Pressure Stall Information, but PSI thresholds should be finetuned per-system. NixOS also doesn't expose PSI configuration

Running `systemd-oomd` is counterproductive for `mglru`.
In a case where `systemd-oomd` kicks in before `mglru` you might lose an entire process group instead of a singular process, whereas `mglru` woud still be likely to have saved you, especially with `min_ttl_ms` higher than `0` (noted though: `min_ttl_ms` currently also isn't exposed via nixOS, needs to be set via a systemd unit file).
Conversely, if `mglru` kills a process before `systemd-oomd` (unlikely, but who knows), you.. prevented out-of-memory slightly more aggressively.

**Comment by K900** [comment]
MGLRU is not an OOM killer. I have no idea where you got this from.

**Comment by ahydronous** [comment]
https://docs.kernel.org/admin-guide/mm/multigen_lru.html

> Users can write N to min_ttl_ms to prevent the working set of N milliseconds from getting evicted. The OOM killer is triggered if this working set cannot be kept in memory. In other words, this option works as an adjustable pressure relief valve, and when open, it terminates applications that are hopefully not being used.

Technically its a feature built on top of MLGRU + OOM killer, but I felt like I didn't need to be pedantic. Ah well.

**Comment by K900** [comment]
Anyway, the whole purpose of oomd is to trigger _before_ the kernel OOM killer, because the kernel OOM killer triggers at the last possible moment, which is not something you want on most systems. Also, our default config doesn't even enable oomd to take action against units, so your entire point is moot. That said, we're likely going to enable it by default eventually, and this will still be a good idea.

**Comment by ahydronous** [comment]
> because the kernel OOM killer triggers at the last possible moment,

That's the point of `min_ttl_ms`, making that tunable. The kernel documentation points to `1000` being a good base value (they refer to <100ms jank being unnoticeable by humans, so perhaps 1 = 0.1 millisecond), and give the suggestion of `3000` as another value that will reduce jank at the cost of more premature OOM kills.
That's also why I brought this up. 

I'm not sure at what value OOM kills would be at the same memory pressure that `systemd-oomd` chooses to evict, but if you'd want consistency in experience you could just set it to that. And now you're no longer killing process groups :)

**Comment by K900** [comment]
Killing process groups is generally _good_, and `min_ttl_ms` will not kick in until you're already _very_ close to out of memory, no matter how high you set that value.

**Comment by mweinelt** [comment]
None of the slices is enabled by default. You would still need to opt into using oomd as it stands.

https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/boot/systemd/oomd.nix#L15-L17

**Comment by ahydronous** [comment]
> killing process groups is generally good

Is it?
Two examples that immediately jump to mind are a browser and shells. If for some reason the browser doesn't catch a tab going crazy, `oomd` could kill your browser process group taking all tabs with it. And any shell action that does `fork()` without `setpgid` / `setsid` (tbh, not sure how rare that is) means that anything in the process group of the parent shell gets nuked.

Thank you for the extra info btw mweinelt! Does this mean that `oomd` is currently doing nothing and basically spinning its wheels?

**Comment by ahydronous** [comment]
I did some more research because I figured maybe I was just being stupid, and perhaps more people will find this issue via search.

Good description of ideas and pros behind `oomd`:
https://engineering.fb.com/2018/07/19/production-engineering/oomd/

One of the top results about `systemd-oomd`, with a lot of complaints (but `oom-killer` also has mountains of complaints): https://www.reddit.com/r/Fedora/comments/10s06fd/why_is_systemdoomd_still_a_thing/.

Given that everything works well, it would seem that `oomd` would be the better choice as it has more information in userspace. But apparently `oomd` can be quite aggressive.

Perhaps the truth is somewhere in the middle, with a downtuned `oomd` and `min_ttl_mms` tuned to a very low value (`100`)? so it only kicks in at a time where even `oomd` is stalling.

At any rate, apologies for coming on too strongly.
